feat: add possibility to cross change data between services

marianfoo opened the pull request:
This PR aims to support cross-service updates allowing updating data in secondService from a action in firstService 

## Changes

Introducing the Service Registry that contains a map of services, so i introduce the saving and discovery of other services without changing something in the core logic.  
In `dataAccess.ts`is a new method is introduced to get the other Service from the registry. 
In 

Added tests
Added and updated documentation
Added test sample with a new app

Is this in general the right approach?

nlunets commented:
In general this would be the right approach, thanks for the start of contribution !
I would however slightly change a few things

- The test just mocks everything, currently all the test at this level rather work by instantiating proper objects, so i would keep it this way. The feature is not complex enough to warrant testing in isolation
- Finding the service using the url is a bit of a mouthful sometimes or could be error prone if your services are close, name wise. I would at least add the option to define an alias for each service to be able to refer to it more easily
- I don't know if this should be a new method or an extension to the getEntityInterface to specify where you want to look for it, something like this.base.getEntityInterface("/MyEntity", "service2")

marianfoo commented:
> * The test just mocks everything, currently all the test at this level rather work by instantiating proper objects, so i would keep it this way. The feature is not complex enough to warrant testing in isolation

changed it to use the files

> * Finding the service using the url is a bit of a mouthful sometimes or could be error prone if your services are close, name wise. I would at least add the option to define an alias for each service to be able to refer to it more easily

Thats a good idea! I added the optional alias parameter, by default you can find the service by url, but also by alias, if you define one

> * I don't know if this should be a new method or an extension to the getEntityInterface to specify where you want to look for it, something like this.base.getEntityInterface("/MyEntity", "service2")

That makes sense to reuse it, it will find the service either bei url or alias now

nlunets commented on docs/MockserverAPI.md:
```diff
@@ -150,13 +166,90 @@ Remove an entry from the data set
+
+- `keyValues` - The key values to identify the entries to fetch
+- Returns array of matching entries
+
+#### hasEntry
+
+Check if an entry with the specified key values exists in the data set.
+
+`hasEntry: (keyValues: KeyDefinitions, odataRequest: ODataRequest) => boolean;`
+
+#### hasEntries
+
+Check if the entity set has any entries.
+
+`hasEntries: (odataRequest: ODataRequest) => boolean;`
+
+#### getAllEntries
+
+Retrieve all entries from the data set.
+
+`getAllEntries: (odataRequest: ODataRequest) => Promise<object[]>;`
+
+#### getEmptyObject
+
+Get an empty object template for the entity type with all properties initialized to default values.
+
+`getEmptyObject: (odataRequest: ODataRequest) => object;`
+
+#### getDefaultElement
+
+Get a default element with generated values for the entity type.
+
+`getDefaultElement: (odataRequest: ODataRequest) => object;`
+
+#### getParentEntityInterface
+
+Retrieve the mockdata entity interface for the parent entity set (useful for draft scenarios).
+
+`getParentEntityInterface: () => Promise<FileBasedMockData | undefined>;`
+
```
Those are more APIs that if you touch them you might break things, so that's why they are not documented initially :)
Might be better to keep them not here for now :)

marianfoo replied:
all good :)

nlunets replied:
maybe as another step :)

nlunets commented on packages/fe-mockserver-core/src/mockdata/functionBasedMockData.ts:
```diff
@@ -161,7 +161,32 @@ export class FunctionBasedMockData extends FileBasedMockData {
+                }
+
+                // If this is a cross-service call, enhance with smart updateEntry behavior
+                if (serviceNameOrAlias) {
```
What's this part for?

marianfoo replied:
when i updated a entry this way it did not update for example only the description but the whole entity
so, when i gave `{"description": "TETS"}` as a parameter, the logic updated the description and because i did not mention all the other parameter they got nulled
as this logic is not in the updateEntry but in the base.updateEntry i added the same logic there
you know a better way?

nlunets replied:
Oh, i see, so this is more or less an existing limitation of the getEntityInterface that you're removing here :)

nlunets replied:
yep looks good

nlunets reviewed (commented):
Overall looks good but a few minor comments

marianfoo reviewed (commented):
.

nlunets commented on packages/fe-mockserver-core/src/data/entitySets/entitySet.ts:
```diff
@@ -244,7 +244,7 @@ export class MockDataEntitySet implements EntitySetInterface {
     protected entitySetDefinition: EntitySet | null;
     protected entityTypeDefinition: EntityType;
-    protected dataAccess: DataAccessInterface;
+    public dataAccess: DataAccessInterface;
```
i changed a bit the logic, should work now

nlunets commented:
I might need to push that onto a branch here, if that's fine with you :)
This way i get sonar running and doing the final checks

nlunets commented:
was merged through another change :)
Thanks @marianfoo
